### PR TITLE
[To rel/1.0] [IOTDB-4978] Fix deadLock caused by blocked operation in IntoOperator

### PR DESCRIPTION
**Pull request by liuminghui233**
see [IOTDB-4978](https://issues.apache.org/jira/browse/IOTDB-4978).

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/AbstractIntoOperator.java [hunk 1]
```diff
@@ -100,27 +138,59 @@ protected void insertMultiTabletsInternally(boolean needCheck) {
...
+    if (writeOperationFuture == null) {
+      return true;
     }
 
-    for (InsertTabletStatementGenerator generator : insertTabletStatementGenerators) {
-      generator.reset();
+    try {
+      if (!writeOperationFuture.isDone()) {
+        return false;
+      }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/AbstractIntoOperator.java:169]: if writeOperationFuture is not done, we should never call next method of into operator, so here if you just want to do a safety check, you should throw a IIlegalStateException.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/AbstractIntoOperator.java [hunk 2]
```diff
@@ -162,42 +232,62 @@ public OperatorContext getOperatorContext() {
...
+    ListenableFuture<?> childBlocked = child.isBlocked();
+    if (writeOperationDone() && childBlocked.isDone()) {
+      return NOT_BLOCKED;
+    } else if (!writeOperationDone() && childBlocked.isDone()) {
+      return writeOperationFuture;
+    } else if (writeOperationDone() && !childBlocked.isDone()) {
+      return childBlocked;
+    } else {
+      return successfulAsList(Arrays.asList(writeOperationFuture, childBlocked));
+    }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/AbstractIntoOperator.java:254]:
```suggestion
    ListenableFuture<?> childBlocked = child.isBlocked();
    boolean writeDone = writeOperationDone();
    if (writeDone && childBlocked.isDone()) {
      return NOT_BLOCKED;
    } else if (childBlocked.isDone()) {
      return writeOperationFuture;
    } else if (writeDone) {
      return childBlocked;
    } else {
      return successfulAsList(Arrays.asList(writeOperationFuture, childBlocked));
    }
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/IntoOperator.java [hunk 3]
```diff
@@ -48,40 +49,70 @@ public IntoOperator(
...
+      if (insertMultiTabletsInternally(false)) {
+        return null;
+      }
+
+      finished = true;
       return constructResultTsBlock();
     }
   }
 
+  /** Return true if write task is submitted during processing */
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/IntoOperator.java:95]: Is this java doc right?

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/AbstractIntoOperator.java (hunk 1 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/AbstractIntoOperator.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/AbstractIntoOperator.java (hunk 2 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/AbstractIntoOperator.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/IntoOperator.java (hunk 3 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/IntoOperator.java]: fixed

**Pull request merged**